### ISSUE TITLE
Plugin health status

### ISSUE THREAD

**Issue by patrick-east** [moby/moby#32797]
There isn't really a good way for plugins to show status to an admin. As-is a plugin has two options when it is in a bad state. It can exit, and maybe be restarted or marked as disabled depending on when. Alternatively it can just fail any requests that come into it. Both of them make for somewhat mysterious errors being encountered by a user/admin that requires looking at the logs to see what is happening.

With that in mind, I would like to propose something like a status field of plugins along with some way for them to report their state.

Over in https://github.com/moby/moby/issues/31923 the CSI proposes a "probe" API for volume plugins, but something like that might be useful for any type of plugin.

Alternatively (or maybe in addition?) just being able to use something like the "HEALTHCHECK" would work too.

A few questions for the community then to get things moving:
* Is this a feature that would be useful for all plugins? Or just limited to the CSI probe functionality in volume plugins?
* Can we get something like HEALTHCHECK and have it be useful? Or does it just add new problems?
* Are there any other features/functionality already planned that would help with this?

**Comment by patrick-east** [comment]
For some context, this was something discussed in the hallway at DockerCon last week with @tiborvass, @msmith491, and @taherv and wanted to get more folks involved.

**Comment by cpuguy83** [comment]
IIRC, this is not what CSI's `probe` endpoint is for.

I'm not sure this is needed because when a user attempts to use a plugin (by specifying it as a driver), if it is in a bad state it should report the error at that point.
In addition, the stdout/stderr of the plugin is piped to docker's stdout/stderr, so it can send logs to the main dockerd log for troubleshooting.

**Comment by patrick-east** [comment]
I agree that its not 100% what the `probe` api is for, but its not too far off, and the end result is not much different. Looking at the doc:

> Caller can use this to probe which machines can support specific storage plugins and schedule workloads accordingly

Which is one side effect that would be ideal with some idea of a health check. Having the plugin running versus having the plugin be able to actually handle requests is not the same, but there is no distinction today. Thats really the main motivation for this. Adding in the smarts to check if some particular type of request is supported makes it even better.

IMO this is more like a quality of life enhancement for troubleshooting. You are correct that an error would be reported when they try to use it... but if you are monitoring the infrastructure wouldn't it be better to know something is wrong *before* it gets used and breaks?

**Comment by cpuguy83** [comment]
@patrick-east But the plugin libraries already handle this.
There's an activation sequence in place for just this reason, with exponential backoff, etc.

Do not think a status field is a very good thing for monitoring.
Metrics to track error rates may be a better approach.

**Comment by patrick-east** [comment]
Good points. Maybe more of a philosophical debate on the monitoring use-case :smile: 

Just to play devils advocate a little bit... Why was a health check and status introduced for normal containers? Bonus topic, why are we not treating plugin containers the same as app containers?

**Comment by cpuguy83** [comment]
@patrick-east It's a different case, and ideally plugins are as simple as possible... ie, it's not working then it should exit.
Plugins don't get treated the same because they are different things that happen to have similar runtime implementations.

Not completely opposed to adding healthchecks... but it's a lot of complexity for something that should remain simple.

**Comment by patrick-east** [comment]
Yea, I agree, Ideally they are simple. What a few of us are finding though is that it's not really that way for some of these.  For some context...

If I were to describe an application that was a service which processes REST API requests on the front end, and on the back end relied on RPC calls to several other daemons/services on the system, at least one remote system, and potentially required access to persistent storage to share state across a cluster of these API handler services... Then to make things worse, if the dependencies on the local daemons was pretty strict for versioning, enough that it made sense to bundle them together with the API service as a group of containers... the application sounds not so simple, right? But nothing so crazy a simple compose file couldn't describe?

Well... that's kind of what we're dealing with with volume plugins that work with iscsid, multipathd, some remote storage backend(s), and need to share state between each member of the cluster. To make them compatible across different systems we start to need pretty complicated management systems.

I feel like maybe I'm getting a bit off topic for the health checks, but I guess its a pretty good first step towards treating plugins more like apps.
